engine/src/actions/movement.rs

```diff
@@ -36,13 +38,38 @@ pub fn down(entity_id: i32) -> ActionData {
     }
 }
 
-pub fn process(world: &mut World, action: ActionData) {
-    if let Some(entity) = world.get_entity(action.entity_id) {
-        if let Some(dir) = action.direction {
-            let coord = operate(entity.coord, dir);
-            let new_entity = entity.with_coordinate(coord);
-            world.update_entity(new_entity);
-        }
+pub fn process(world: &mut World, action: ActionData) -> EngineResult<()> {
+    let entity = world.get_entity(action.entity_id)?;
+
+    if let Some(dir) = action.direction {
+        let new_coord = operate(entity.coord, dir);
+
+        is_inside_world(world, new_coord)?;
+        is_position_available(world, new_coord)?;
+
+        let new_entity = entity.with_coordinate(new_coord);
+        world.update_entity(new_entity);
+    }
+
+    Ok(())
+}
+
+fn is_inside_world(world: &World, coord: Coordinate) -> EngineResult<()> {
+    if coord.x < 0.0
+        || coord.y < 0.0
+        || coord.x >= world.size_x as f32
+        || coord.y >= world.size_y as f32
```

Comment:
This comparison made me wonder if we should represent the edges of the world as coordinates instead of having `size_x` and `size_y`. Also, if we need to use floats for interactions with ggez, shouldn't we use them for these fields as well?

Reply:
BTW, expanding on the coordinates idea above. If we did that, we could have a generic function over coordinates that asks whether `this` coordinates `is_within` two other coordinates given as input.

Reply:
:+1 on the coordinates.